### PR TITLE
[QNN] Implement quantized avg_pool2d

### PR DESCRIPTION
**Pull request by kparzysz-quic**
* qnn.avg_pool2d is integrated into Relay and integration tests are added
* FQ2I is modified to pick up qnn.avg_pool2d
* Canonicalization and layout conversion for qnn.avg_pool2d are implemented
* Dynamic PoolArea computation is implemented for both qnn and slice_op avg_pool2d to exclude padding area in averaging calculation when count_include_pad=False

### CODE REVIEW

File: python/tvm/topi/hexagon/utils.py [hunk 1]
```diff
@@ -24,10 +24,30 @@
...
+    return get_const_int(expr)
+
+
+def get_const_float_value(expr):
+    if isinstance(expr, te.Tensor):
+        assert isinstance(expr.op.body[0], tir.FloatImm)
+        return expr.op.body[0].value
+    return get_const_float(expr)
+
+
```
Comment (ibsidorenko) [inline python/tvm/topi/hexagon/utils.py:50]: `is_scalar/get_const_int_value/get_const_float_value` - it looks like code duplication with `python/tvm/topi/hexagon/qnn/nn.py` (except for Relax const support).
Agree, `utils.py` is a better place for that. Can we add support for Relax constants here and remove it from `nn.py`?

**Comment by kparzysz-quic** [comment]
Thanks Masa!  I'm planning to address your comments this weekend.

File: src/relay/op/nn/pooling_common.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relay/op/nn/pooling_common.h
+ * \brief Properties def of pooling operator for sharing.
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply src/relay/op/nn/pooling_common.h:22]: Done

File: src/relay/qnn/op/avg_pool2d.cc [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+/*
+ * \brief Forward rewrite the qnn Avg_pool2d op.
+ * \param attrs The QNN Avg_pool2d attrs.
+ * \param new_args The new mutated args to the call node.
+ * \param arg_types The types of input and output.
+ * \return The sequence of Relay ops for qnn Avg_pool2d op.
+ * \note Lowering of the qnn.Avg_pool2d operator
+
+ *  Quantized Avg_pool2d will take one quantized input tensor and returns another
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply src/relay/qnn/op/avg_pool2d.cc:152]: Done

File: src/relay/qnn/op/avg_pool2d.cc [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  return InferCorrectLayoutOutput(input_layouts, output_layouts, attrs);
+}
+
+/*
+ * \brief Forward rewrite the qnn Avg_pool2d op.
+ * \param attrs The QNN Avg_pool2d attrs.
+ * \param new_args The new mutated args to the call node.
+ * \param arg_types The types of input and output.
+ * \return The sequence of Relay ops for qnn Avg_pool2d op.
+ * \note Lowering of the qnn.Avg_pool2d operator
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply src/relay/qnn/op/avg_pool2d.cc:150]: Done

File: src/relay/qnn/op/avg_pool2d.cc [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+  Layout channel_layout = Layout("C");
+  Array<Layout> input_layouts = {avgpool_new_layouts->input_layouts[0], channel_layout,
+                                 channel_layout, channel_layout, channel_layout};
+  Array<Layout> output_layouts = avgpool_new_layouts->output_layouts;
+  return InferCorrectLayoutOutput(input_layouts, output_layouts, attrs);
+}
+
+/*
+ * \brief Forward rewrite the qnn Avg_pool2d op.
+ * \param attrs The QNN Avg_pool2d attrs.
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply src/relay/qnn/op/avg_pool2d.cc:146]: Done

File: src/relay/qnn/op/avg_pool2d.cc [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relay/qnn/op/Avg_pool2d.cc
+ * \brief Property def of qnn Avg_pool2d operator.
+ */
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply src/relay/qnn/op/avg_pool2d.cc:23]: Done

File: python/tvm/topi/hexagon/utils.py (hunk 1 above)
Comment (ibsidorenko) [reply python/tvm/topi/hexagon/utils.py:50]: oh.. I've confused with unity branch. In main branch `is_scalar/get_const_int_value/get_const_float_value`  are identical with `python/tvm/topi/hexagon/qnn/nn.py`

File: tests/python/contrib/test_hexagon/test_qnn_op_integration.py [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+""" Test Relay integrated qnn ops
+There are two types of tests for qnn ops in this file. One to verify the
+correctness of the relay integration and the other one to verify
+the fake quantization to integer implemented for picking up the qnn op.
+The former is only executed when qnn canonicalization is disabled.
+The latter is executed both with and without canonicalization.
+"""
+# TODO: We might want to distribute these test cases into other test cases such as
+# test_wo_qnn_canonicalization and test_pass_fake_quantization_to_integer in the future.
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply tests/python/contrib/test_hexagon/test_qnn_op_integration.py:27]: There is already a test for avgpool2d in `tests/python/relay/test_pass_fake_quantization_to_integer.py`.  The tests from this file (`test_qnn_op_integration.py`) execute on Hexagon and require all the fixtures from `tests/python/contrib/test_hexagon`.

Regarding migration of `test_wo_qnn_canonicalization` to this file: I've started working on this, but this file contains a number of fixtures that assume that there is only one tensor input.  Seems like migrating tests from this file over to "wo_canonicalization" would be easier (both to write and to read later on).  What do you think?

File: src/relay/qnn/op/avg_pool2d.cc [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+InferCorrectLayoutOutput QnnAvgPoolInferCorrectLayout(const Attrs& attrs,
+                                                      const Array<Layout>& new_in_layouts,
+                                                      const Array<Layout>& old_in_layouts,
+                                                      const Array<tvm::relay::Type>& old_in_types) {
+  // Use Relay AvgPool2D Infer correct layout.
+  auto avgpool_new_layouts =
+      PoolInferCorrectLayout<AvgPool2DAttrs>(attrs, new_in_layouts, old_in_layouts, old_in_types);
+
+  // Fill the layouts of remaining input tensors - scales and zero points. The layouts of these
+  // tensors can be treated as channel layout.
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply src/relay/qnn/op/avg_pool2d.cc:136]: Seems to work ok with "undef".  Done.

File: python/tvm/topi/hexagon/slice_ops/avg_pool2d.py [hunk 9]
```diff
@@ -16,118 +16,206 @@
...
+        n, h, w, c = s.get_loops(Avg)
+    else:
+        n, c, h, w = s.get_loops(Avg)
+    _, ci = s.split(c, [None, 1024])
     cio, cii = s.split(ci, [None, 64])
     s.vectorize(cii)
 
     # Schedule 'Sum'
-    s.compute_at(Sum, cio)
+    # s.compute_at(Sum, cio)
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply python/tvm/topi/hexagon/slice_ops/avg_pool2d.py:218]: Done.

File: tests/python/contrib/test_hexagon/test_qnn_op_integration.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (masahi) [reply tests/python/contrib/test_hexagon/test_qnn_op_integration.py:27]: > There is already a test for avgpool2d

ah ok I was confused about the FQ2I change and the qnn op addition.

>  Seems like migrating tests from this file over to "wo_canonicalization"

Sounds good, I think you can also rename that file to `test_qnn_op_integration.py`

[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply tests/python/contrib/test_hexagon/test_qnn_op_integration.py]: Done

**Comment by kparzysz-quic** [comment]
Rebased to fix the GPU build error.

**Pull request merged**